### PR TITLE
feat: add ability to set nonce and tenant id

### PR DESCRIPTION
**Pull request by love-bhardwaj**
This gives the ability to set these values on the stratgey for each auth request individually

### CODE REVIEW

File: tests/src/strategy.spec.ts [hunk 1]
```diff
@@ -3,16 +3,61 @@
...
+      authorizationURL: testData.authorizationURL,
+      tokenURL: testData.tokenURL,
+      clientID: testData.clientID,
+      clientSecret: testData.clientSecret,
+      callbackURL: testData.callbackURL,
+      scope: testData.scope,
+      scopeSeparator: ' ',
+      state: true,
+      passReqToCallback: true,
+      nonce: testData.nonce,
```
Comment (mitechie) [inline tests/src/strategy.spec.ts:38]: should the nonce value still be here as it's required per-request and different for each per the spec?

File: tests/src/strategy.spec.ts [hunk 2]
```diff
@@ -34,6 +79,7 @@ describe('Strategy with passport', () => {
     // It only provides a meaning when served with a 3xx (redirection) or 201 (created) status response.
     expect(response.statusCode).toBe(302);
     const redirectURL = new URL(response.header.location);
+    console.log(response.headers.location);
```
Comment (clwiseman) [inline tests/src/strategy.spec.ts:82]: extra console.log

File: tests/src/strategy.spec.ts (hunk 1 above)
Comment (love-bhardwaj) [reply tests/src/strategy.spec.ts:38]: POC 🙂 will remove once this method is feasible for all

File: tests/src/strategy.spec.ts (hunk 2 above)
Comment (love-bhardwaj) [reply tests/src/strategy.spec.ts:82]: POC 🙂 will remove once this methods is feasible for all

File: src/strategy.ts [hunk 3]
```diff
@@ -12,23 +12,39 @@ export default class SnykOAuth2Strategy extends Strategy {
...
    * to get profile for their records
    */
-  private _nonce: string;
+  private _nonce = '';
+  private _tenantId = '';
   private _profileFunc?: ProfileFunc;
 
   constructor(options: SnykStrategyOptions, verify: VerifyFunctionWithRequest) {
     super(options, verify);
-    this._nonce = options.nonce;
```
Comment (ildikodencsak) [inline src/strategy.ts:20]: Shouldn't we keep supporting also the nonce from the options?

File: tests/src/strategy.spec.ts [hunk 4]
```diff
@@ -3,16 +3,61 @@
...
 
   // This will trigger the auth flow
-  app.get('/auth', passport.authenticate('snyk-oauth2', { state: 'test' }));
+  app.get(
+    '/auth',
+    (req, res, next) => {
+      snykOAuth2.nonce = testData.nonce;
+      next();
+    },
+    passport.authenticate('snyk-oauth2', { state: 'test' }),
```
Comment (pavel-solodilov) [inline tests/src/strategy.spec.ts:59]: Hmmmm, what if `passport.authenticate` has any async things internally, and between the nonce being retrieved from the `snykOAuth2` instance and `passport.authenticate` being called you have another concurrent request coming in? This screams race condition to me 🤔

Comment (pavel-solodilov) [reply tests/src/strategy.spec.ts:59]: The other solution would be to still pass it into `passport.authenticate`, but either:
- Amend Passport.js to be able to pass this information.
- Export our own wrapper around `passport.authenticate` when we distribute our strategy that accepts the `nonce` value.

Comment (love-bhardwaj) [reply tests/src/strategy.spec.ts:59]: I can see the problem, but I don't see any other of doing it without updating the `passport` code. There is another hackish solution where you can just pass the `nonce` and `tenant_id` values using `state`. Something like `state: `state-value&nonce=${nonce}&tenant_id=${tenantId}`'

File: src/strategy.ts (hunk 3 above)
Comment (love-bhardwaj) [reply src/strategy.ts:20]: Setting it through the constructor means it is the same for all requests going out, which defeats the purpose of `nonce`.

File: tests/src/strategy.spec.ts (hunk 4 above)
Comment (pavel-solodilov) [reply tests/src/strategy.spec.ts:59]: Waaaait, I'm working on an assumption that Passport calls the `authorizationParams` function, but I don't see it doing it o.O

Comment (love-bhardwaj) [reply tests/src/strategy.spec.ts:59]: I think I found a way of doing it. We need to pass on the options we get in the `authorizationParams` as you see the passport code returns an empty object. After doing that the solution you suggested should work.

Comment (pavel-solodilov) [reply tests/src/strategy.spec.ts:59]: There's also the `tokenParams` call: https://github.dev/jaredhanson/passport-oauth2/blob/ee3fe9f17c0f3a90f2d9d938f267e9942b9fba49/lib/strategy.js#L168

Comment (love-bhardwaj) [reply tests/src/strategy.spec.ts:59]: We can forwards option on both 🙂

Comment (love-bhardwaj) [reply tests/src/strategy.spec.ts:59]: Here is the PR https://github.com/snyk/passport-snyk-oauth2/pull/8 that should resolve it. Thanks for the inputs 🙇

**Comment by love-bhardwaj** [comment]
Closing in favour of https://github.com/snyk/passport-snyk-oauth2/pull/8

**Pull request closed without merging**